File: ToucanGraph.py
import Utils

import networkx as nx

class ToucanGraph:
  def __init__(self, graph=None):
    if graph is None:
      self.graph = nx.DiGraph()  # Use directed graph, change to Graph() for undirected
    elif isinstance(graph, nx.DiGraph):
      self.graph = graph.copy()
    else:
      raise TypeError("graph must be an instance of networkx.DiGraph")

    self.levels = []  # List of lists to store nodes at each level
    self.vecdecl_to_nop = {} # Map of original VecDecl to new Vecdecl NOP
    self.max_node_id = 0

  def load(self, file_path: str) -> None:
    """Load a graph from a text file in the described format."""
    with open(file_path, 'r') as file:
      for lineno, line in enumerate(file):
        if lineno == 0:
          # Read the first line for number of edges and nodes
          first_line = line.strip()
          num_edges, num_nodes = map(int, first_line.split())

          nodes_to_add = []
          edges_to_add = []
          invalid_nodes = set()
          continue

        
        parts = line.strip().split()
        node_id = int(parts[0])
        self.max_node_id = max(node_id, self.max_node_id)
        if len(parts) < 3:
          raise ValueError(f"Node {node_id} is missing weight or neighbors.")

        label = parts[1]
        if not label:
          raise ValueError(f"Node {node_id} has an empty label, which is illegal.")

        weight = int(parts[2])

        # Skip invalid nodes
        if weight < 0:
          invalid_nodes.add(node_id)
          continue

        # Collect node information
        nodes_to_add.append((node_id, {
          "label": label,
          "weight": weight
        }))

        # Collect edges information
        if len(parts) > 3:
          neighbors = map(int, parts[3:])
          for neighbor in neighbors:
            edges_to_add.append((node_id, neighbor))

      # Add all nodes to the graph
      self.graph.add_nodes_from(nodes_to_add)

      # Add all edges to the graph
      edge_count = 0
      for source, target in edges_to_add:
        assert(source not in invalid_nodes)
        if target in invalid_nodes:
          continue
        if not self.graph.has_node(source) or not self.graph.has_node(target):
          source_existance = "Exist" if self.graph.has_node(source) else "NonExist"
          target_existance = "Exist" if self.graph.has_node(source) else "NonExist"
          raise ValueError(f"Line {lineno}: Edge from {source} ({source_existance}) to {target} ({target_existance}) refers to non-existent node.")
        self.graph.add_edge(source, target)
        edge_count += 1

      # Assert the number of nodes and edges
      assert self.graph.number_of_nodes() == num_nodes, f"Expected {num_nodes} nodes, found {self.graph.number_of_nodes()}"
      assert edge_count == num_edges, f"Expected {num_edges} edges, found {edge_count}"

  def create_subgraph(self, node_list):
    """Create a subgraph containing only the specified nodes."""
    subgraph = self.graph.subgraph(node_list).copy()
    return ToucanGraph(subgraph)

  def levelize(self):
    """Legalize the graph by assigning level IDs to each node and updating levels list."""
    if not nx.is_directed_acyclic_graph(self.graph):
      raise ValueError("The graph must be a Directed Acyclic Graph (DAG) to levelize.")

    # Compute the levels using topological sorting
    level_mapping = {}
    self.levels = []  # Reset levels
    for level_id, nodes in enumerate(nx.topological_generations(self.graph)):
      self.levels.append(list(nodes))
      for node in nodes:
        level_mapping[node] = level_id

    # Assign level IDs to node attributes
    for node, level_id in level_mapping.items():
      self.graph.nodes[node]["level_id"] = level_id

  def is_levelized(self) -> bool:
    """Check if the graph is levelized by verifying if levels is empty."""
    return bool(self.levels)

  def __str__(self):
    """Return a string representation of the graph."""
    level_info = f" and {len(self.levels)} levels" if self.levels else ""
    return f"Graph with {self.graph.number_of_nodes()} nodes and {self.graph.number_of_edges()} edges{level_info}."

  # def findAllNodesWithTag(self, tagName):
  #   """Find all nodes with a specific tagName and return a dict with tagValue as the key."""
  #   result = {}
  #   for node, attrs in self.graph.nodes(data=True):
  #     tagValue = attrs.get(tagName)
  #     if tagValue != None:
  #       if tagValue not in result:
  #         result[tagValue] = []
  #       result[tagValue].append(node)
  #   return result
  
  def is_acyclic(self) -> bool:
    """Check if the graph is acyclic."""
    return nx.is_directed_acyclic_graph(self.graph)
  
  def expand_VecDecl(self, vecDeclElements: dict):
    vecDecl_node_ids = []
    nodes_to_remove = []

    for node, attrs in self.graph.nodes(data=True):
      tagValue = attrs.get("label")
      if tagValue == "VecDecl":
        node_weight = attrs.get("weight")
        vecDecl_node_ids.append((node, node_weight))
        nodes_to_remove.append(node)
    print(len(vecDecl_node_ids))

    for node, weight in vecDecl_node_ids:
      vec_input_nodes = list(self.graph.predecessors(node))
      vec_user_nodes = list(self.graph.successors(node))

      # Vec info file should be consistant with graph info
      assert(node in vecDeclElements)

      vec_element_op_ids = vecDeclElements[node]

      # At least one vecDecl element
      assert(weight > 0)
      # should also be consistant
      assert(weight == len(vec_element_op_ids))
      assert(len(vec_input_nodes) <= weight)

      nodes_to_add = []
      edges_to_add = []
      new_node_list = []
      for i in range(0, weight):
        # insert NOP
        self.max_node_id += 1
        node_id = self.max_node_id
        assert(node_id not in self.graph.nodes())

        nodes_to_add.append((node_id, {
          "label": "VecDecl_LUT_NOP",
          "weight": 1,
          "original_vec_decl": node
        }))
        new_node_list.append(node_id)

        edge_src = vec_element_op_ids[i]
        assert(edge_src in vec_input_nodes)
        for d in vec_user_nodes:
          edges_to_add.append((edge_src, node_id))
          edges_to_add.append((node_id, d))

      assert(node not in self.vecdecl_to_nop)
      self.vecdecl_to_nop[node] = new_node_list
      self.graph.add_nodes_from(nodes_to_add)
      self.graph.add_edges_from(edges_to_add)
    assert(len(nodes_to_remove) <= len(vecDeclElements))
    self.graph.remove_nodes_from(nodes_to_remove)

  def remove_ConstDecl(self):
    nodes_to_remove = []
    for node, attrs in self.graph.nodes(data=True):
      tagValue = attrs.get("label")
      if tagValue == "ConstDecl":
        nodes_to_remove.append(node)
    print(f"Remove {len(nodes_to_remove)} ConstDecl nodes")
    self.graph.remove_nodes_from(nodes_to_remove)

  def save_vector_def_info(self, filename: str):
    with open(filename, 'w') as out:
      for vecDecl_node, nop_list in self.vecdecl_to_nop.items():
        line = [vecDecl_node]
        line.extend(nop_list)
        assert(len(nop_list) != 0)
        out.write(' '.join(map(lambda x: str(x), line)))
        out.write("\n")




  
  # def merge_nodes(self, nodes_to_merge, new_node_attributes=None):
  #   """Merge multiple nodes into a single node.

  #   Args:
  #     nodes_to_merge: List of node IDs to be merged.
  #     new_node_attributes: Attributes for the new node (optional).

  #   Returns:
  #     new_node_id: The ID of the newly created node.
  #   """
  #   if not all(node in self.graph for node in nodes_to_merge):
  #     raise ValueError("All nodes to merge must exist in the graph.")

  #   # Generate a new node ID
  #   new_node_id = max(self.graph.nodes) + 1

  #   # Collect edges for the new node
  #   incoming_edges = set()
  #   outgoing_edges = set()

  #   for node in nodes_to_merge:
  #     incoming_edges.update((pred, new_node_id) for pred in self.graph.predecessors(node) if pred not in nodes_to_merge)
  #     outgoing_edges.update((new_node_id, succ) for succ in self.graph.successors(node) if succ not in nodes_to_merge)

  #   # Remove the old nodes
  #   self.graph.remove_nodes_from(nodes_to_merge)

  #   # Add the new merged node
  #   self.graph.add_node(new_node_id, **(new_node_attributes or {}))

  #   # Add the collected edges
  #   self.graph.add_edges_from(incoming_edges)
  #   self.graph.add_edges_from(outgoing_edges)

  #   return new_node_id
  
  # def dropAllConstDecls

if __name__ == "__main__":
  # test
  fileName = "design_before_cut.graph"


  g = ToucanGraph()
  g.load(fileName)
  g.levelize()
  assert(g.is_acyclic())

  print(g)

  print(len(g.levels))
  for node in g.levels[-1]:
    print(g.graph.nodes[node].get('label'))

  Utils.print_memory_usage()

File: src/ToucanGraph.cpp
#include "ToucanGraph.h"
#include "Utils.h"
#include <fstream>
#include <sstream>
#include <iostream>
#include <algorithm>
#include <queue>
#include <cassert>
#include <functional>
#include <stdexcept>

// NodeTag utility functions implementation
NodeTag string_to_node_tag(const std::string& str) {
    if (str == "ConstDecl") return NodeTag::ConstDecl;
    if (str == "RegRead") return NodeTag::RegRead;
    if (str == "MemRead") return NodeTag::MemRead;
    if (str == "VecDecl") return NodeTag::VecDecl;
    if (str == "VecDecl_LUT_NOP") return NodeTag::VecDecl_LUT_NOP;
    if (str == "VecRead") return NodeTag::VecRead;
    if (str == "LUT") return NodeTag::LUT;
    if (str == "VecArith") return NodeTag::VecArith;
    if (str == "VecLogic") return NodeTag::VecLogic;
    if (str == "Print") return NodeTag::Print;
    if (str == "Stop") return NodeTag::Stop;
    if (str == "RegWrite") return NodeTag::RegWrite;
    if (str == "MemWrite") return NodeTag::MemWrite;
    return NodeTag::UNKNOWN;
}

const char* node_tag_to_string(NodeTag tag) {
    switch (tag) {
        case NodeTag::ConstDecl: return "ConstDecl";
        case NodeTag::RegRead: return "RegRead";
        case NodeTag::MemRead: return "MemRead";
        case NodeTag::VecDecl: return "VecDecl";
        case NodeTag::VecDecl_LUT_NOP: return "VecDecl_LUT_NOP";
        case NodeTag::VecRead: return "VecRead";
        case NodeTag::LUT: return "LUT";
        case NodeTag::VecArith: return "VecArith";
        case NodeTag::VecLogic: return "VecLogic";
        case NodeTag::Print: return "Print";
        case NodeTag::Stop: return "Stop";
        case NodeTag::RegWrite: return "RegWrite";
        case NodeTag::MemWrite: return "MemWrite";
        case NodeTag::UNKNOWN: return "UNKNOWN";
    }
    return "UNKNOWN";
}

bool is_valid_node_tag(NodeTag tag) {
    return tag != NodeTag::UNKNOWN;
}

bool is_exclude_node_tag(NodeTag tag) {
    switch (tag) {
        case NodeTag::ConstDecl:
        case NodeTag::RegRead:
        case NodeTag::MemRead:
        case NodeTag::VecDecl:
        case NodeTag::VecRead:
        case NodeTag::VecArith:
        case NodeTag::VecLogic:
        case NodeTag::Print:
        case NodeTag::Stop:
        case NodeTag::RegWrite:
        case NodeTag::MemWrite:
            return true;
        default:
            return false;
    }
}

bool is_merge_result_node_tag(NodeTag tag) {
    return tag == NodeTag::VecArith || tag == NodeTag::VecRead;
}

ToucanGraph::ToucanGraph() : max_node_id(0), edge_count(0) {}

void ToucanGraph::load(const std::string& file_path) {
    std::ifstream file(file_path);
    if (!file.is_open()) {
        throw std::runtime_error("Cannot open file: " + file_path);
    }

    std::string line;
    int lineno = 0;
    int num_edges = 0, num_nodes = 0;
    std::vector<std::pair<int, int>> edges_to_add;
    std::unordered_set<int> invalid_nodes;

    while (std::getline(file, line)) {
        if (lineno == 0) {
            // Read the first line for number of edges and nodes
            std::istringstream iss(line);
            iss >> num_edges >> num_nodes;
            lineno++;
            continue;
        }

        std::vector<std::string> parts = Utils::split(line, ' ');
        if (parts.empty()) continue;

        int node_id = std::stoi(parts[0]);
        max_node_id = std::max(node_id, max_node_id);

        if (parts.size() < 3) {
            throw std::runtime_error("Node " + std::to_string(node_id) + " is missing weight or neighbors.");
        }

        std::string label = parts[1];
        if (label.empty()) {
            throw std::runtime_error("Node " + std::to_string(node_id) + " has an empty label, which is illegal.");
        }

        int weight = std::stoi(parts[2]);

        // Skip invalid nodes
        if (weight < 0) {
            invalid_nodes.insert(node_id);
            lineno++;
            continue;
        }

        // Convert string label to NodeTag enum
        NodeTag tag = string_to_node_tag(label);
        if (tag == NodeTag::UNKNOWN) {
            throw std::runtime_error("Unknown node tag/label '" + label + "' at node " + std::to_string(node_id));
        }

        // Add node
        NodeAttributes attrs;
        attrs.tag = tag;
        attrs.weight = weight;
        add_node(node_id, attrs);

        // Collect edges
        for (size_t i = 3; i < parts.size(); ++i) {
            int neighbor = std::stoi(parts[i]);
            edges_to_add.emplace_back(node_id, neighbor);
        }

        lineno++;
    }

    // Add all edges
    int actual_edge_count = 0;
    for (const auto& edge : edges_to_add) {
        int source = edge.first, target = edge.second;
        
        if (invalid_nodes.contains(source) || invalid_nodes.contains(target)) {
            continue;
        }
        
        if (!has_node(source) || !has_node(target)) {
            throw std::runtime_error("Edge from " + std::to_string(source) + 
                                   " to " + std::to_string(target) + " refers to non-existent node.");
        }
        
        add_edge(source, target);
        actual_edge_count++;
    }

    // Verify counts
    if (nodes.size() != static_cast<size_t>(num_nodes)) {
        throw std::runtime_error("Expected " + std::to_string(num_nodes) + 
                               " nodes, found " + std::to_string(nodes.size()));
    }
    if (actual_edge_count != num_edges) {
        throw std::runtime_error("Expected " + std::to_string(num_edges) + 
                               " edges, found " + std::to_string(actual_edge_count));
    }
}

void ToucanGraph::levelize() {
    if (!is_acyclic()) {
        throw std::runtime_error("The graph must be a Directed Acyclic Graph (DAG) to levelize.");
    }

    levels.clear();
    
    // Calculate in-degrees
    std::unordered_map<int, int> in_degree;
    for (const auto& node : nodes) {
        in_degree[node.first] = get_in_degree(node.first);
    }

    // Topological sort using Kahn's algorithm
    std::queue<int> queue;
    for (const auto& pair : in_degree) {
        if (pair.second == 0) {
            queue.push(pair.first);
        }
    }

    int current_level = 0;
    while (!queue.empty()) {
        int level_size = queue.size();
        std::vector<int> current_level_nodes;
        
        for (int i = 0; i < level_size; ++i) {
            int node = queue.front();
            queue.pop();
            
            current_level_nodes.push_back(node);
            nodes[node].level_id = current_level;
            
            // Reduce in-degree of successors
            for (int successor : get_successors(node)) {
                in_degree[successor]--;
                if (in_degree[successor] == 0) {
                    queue.push(successor);
                }
            }
        }
        
        levels.push_back(current_level_nodes);
        current_level++;
    }
}

bool ToucanGraph::is_acyclic() const {
    // Use DFS to detect cycles
    std::unordered_set<int> white, gray, black;
    
    for (const auto& node : nodes) {
        white.insert(node.first);
    }
    
    std::function<bool(int)> dfs = [&](int node) -> bool {
        white.erase(node);
        gray.insert(node);
        
        for (int successor : get_successors(node)) {
            if (gray.count(successor)) {
                return false; // Back edge found, cycle detected
            }
            if (white.count(successor) && !dfs(successor)) {
                return false;
            }
        }
        
        gray.erase(node);
        black.insert(node);
        return true;
    };
    
    while (!white.empty()) {
        int start = *white.begin();
        if (!dfs(start)) {
            return false;
        }
    }
    
    return true;
}

void ToucanGraph::expand_VecDecl(const std::unordered_map<int, std::vector<int>>& vecDeclElements) {
    std::vector<int> vecDecl_node_ids;
    std::vector<int> nodes_to_remove;
    int num_new_nodes = 0;

    // Find all VecDecl nodes
    for (const auto& node : nodes) {
        if (node.second.tag == NodeTag::VecDecl) {
            vecDecl_node_ids.push_back(node.first);
            nodes_to_remove.push_back(node.first);
        }
    }

    for (int node : vecDecl_node_ids) {
        std::vector<int> vec_input_nodes = get_predecessors(node);
        std::vector<int> vec_user_nodes = get_successors(node);
        int weight = nodes[node].weight;

        // Vec info file should be consistent with graph info
        auto it = vecDeclElements.find(node);
        assert(it != vecDeclElements.end());

        const std::vector<int>& vec_element_op_ids = it->second;

        // At least one vecDecl element
        assert(weight > 0);
        int nop_node_weight = weight / vec_element_op_ids.size();
        assert(vec_input_nodes.size() <= static_cast<size_t>(weight));

        std::vector<int> new_node_list;
        for (size_t i = 0; i < vec_element_op_ids.size(); ++i) {
            // Insert NOP
            max_node_id++;
            int node_id = max_node_id;
            assert(!has_node(node_id));

            NodeAttributes attrs;
            attrs.tag = NodeTag::VecDecl_LUT_NOP;
            attrs.weight = nop_node_weight;
            attrs.original_vec_decl = node;
            add_node(node_id, attrs);
            new_node_list.push_back(node_id);

            int edge_src = vec_element_op_ids[i];
            assert(std::find(vec_input_nodes.begin(), vec_input_nodes.end(), edge_src) != vec_input_nodes.end());
            add_edge(edge_src, node_id);
            
            for (int d : vec_user_nodes) {
                add_edge(node_id, d);
            }
        }

        assert(vecdecl_to_nop.find(node) == vecdecl_to_nop.end());
        vecdecl_to_nop[node] = new_node_list;
        num_new_nodes += new_node_list.size();

        // Verify in-degree
        for ([[maybe_unused]] int new_node : new_node_list) {
            assert(get_in_degree(new_node) == 1);
        }
    }

    assert(nodes_to_remove.size() <= vecDeclElements.size());
    remove_nodes(nodes_to_remove);
    std::cout << "Expand " << nodes_to_remove.size() << " vectors, add " << num_new_nodes << " new VecDecl_LUT_NOP nodes\n";
}

void ToucanGraph::remove_ConstDecl() {
    std::vector<int> nodes_to_remove;
    for (const auto& node : nodes) {
        if (node.second.tag == NodeTag::ConstDecl) {
            nodes_to_remove.push_back(node.first);
        }
    }
    std::cout << "Remove " << nodes_to_remove.size() << " ConstDecl nodes\n";
    remove_nodes(nodes_to_remove);
}

void ToucanGraph::save_vector_def_info(const std::string& filename) const {
    std::ofstream out(filename);
    if (!out.is_open()) {
        throw std::runtime_error("Cannot open file for writing: " + filename);
    }

    for (const auto& pair : vecdecl_to_nop) {
        out << pair.first;
        for (int nop : pair.second) {
            out << " " << nop;
        }
        out << "\n";
    }
    out.close();
}

std::vector<int> ToucanGraph::get_predecessors(int node) const {
    if (reverse_adjacency_list.contains(node)) {
        return reverse_adjacency_list.at(node);
    }
    return {};
}

std::vector<int> ToucanGraph::get_successors(int node) const {
    if (adjacency_list.contains(node)) {
        return adjacency_list.at(node);
    }
    return {};
}

bool ToucanGraph::has_node(int node) const {
    return nodes.contains(node);
}

bool ToucanGraph::has_edge(int from, int to) const {
    auto it = adjacency_list.find(from);
    if (it != adjacency_list.end()) {
        const auto& successors = it->second;
        return std::find(successors.begin(), successors.end(), to) != successors.end();
    }
    return false;
}

int ToucanGraph::get_in_degree(int node) const {
    auto it = reverse_adjacency_list.find(node);
    return it != reverse_adjacency_list.end() ? it->second.size() : 0;
}

int ToucanGraph::get_out_degree(int node) const {
    auto it = adjacency_list.find(node);
    return it != adjacency_list.end() ? it->second.size() : 0;
}

int ToucanGraph::max_node() const {
    if (nodes.empty()) {
        return -1; // No nodes in graph
    }
    
    int max_id = -1;
    for (const auto& pair : nodes) {
        max_id = std::max(max_id, pair.first);
    }
    return max_id;
}

std::unique_ptr<ToucanGraph> ToucanGraph::create_subgraph(const std::unordered_set<int>& node_list) const {
    auto subgraph = std::make_unique<ToucanGraph>();
    
    // Add nodes
    for (int node : node_list) {
        if (has_node(node)) {
            subgraph->add_node(node, nodes.at(node));
        }
    }
    
    // Add edges
    for (int node : node_list) {
        for (int successor : get_successors(node)) {
            if (node_list.count(successor)) {
                subgraph->add_edge(node, successor);
            }
        }
    }
    
    return subgraph;
}

void ToucanGraph::add_node(int node_id, const NodeAttributes& attrs) {
    nodes[node_id] = attrs;
    adjacency_list[node_id] = {};
    reverse_adjacency_list[node_id] = {};
}

void ToucanGraph::add_edge(int from, int to) {
    adjacency_list[from].push_back(to);
    reverse_adjacency_list[to].push_back(from);
    edge_count++;
}

void ToucanGraph::remove_node(int node_id) {
    if (!has_node(node_id)) return;
    
    // Remove all edges involving this node
    for (int predecessor : get_predecessors(node_id)) {
        auto& successors = adjacency_list[predecessor];
        successors.erase(std::remove(successors.begin(), successors.end(), node_id), successors.end());
        edge_count--;
    }
    
    for (int successor : get_successors(node_id)) {
        auto& predecessors = reverse_adjacency_list[successor];
        predecessors.erase(std::remove(predecessors.begin(), predecessors.end(), node_id), predecessors.end());
    }
    
    // Remove node
    nodes.erase(node_id);
    adjacency_list.erase(node_id);
    reverse_adjacency_list.erase(node_id);
}

void ToucanGraph::remove_nodes(const std::vector<int>& nodes_to_remove) {
    for (int node : nodes_to_remove) {
        remove_node(node);
    }
}
